lib/clients/public.js

```diff
@@ -1,206 +1,213 @@
+  }
+  delete(...args) {
+    return this.request('delete', ...args);
+  }
 
-  prototype.addHeaders = function(obj, additional) {
+  addHeaders(obj, additional) {
     obj.headers = obj.headers || {};
-    return _.assign(obj.headers, {
-      'User-Agent': 'gdax-node-client',
-      'Accept': 'application/json',
-      'Content-Type': 'application/json',
-    }, additional);
-  };
-
-  prototype.makeRelativeURI = function(parts) {
+    return Object.assign(
+      obj.headers,
+      {
+        'User-Agent': 'gdax-node-client',
+        Accept: 'application/json',
+        'Content-Type': 'application/json',
+      },
+      additional
+    );
+  }
+
+  makeRelativeURI(parts) {
     return '/' + parts.join('/');
-  };
+  }
 
-  prototype.makeAbsoluteURI = function(relativeURI) {
+  makeAbsoluteURI(relativeURI) {
     return this.apiURI + relativeURI;
-  };
+  }
 
-  prototype.makeRequestCallback = function(callback) {
+  makeRequestCallback(callback, resolve, reject) {
     return function(err, response, data) {
       try {
         data = JSON.parse(data);
       } catch (e) {
-        data = null
+        data = null;
       }
       callback(err, response, data);
+      if (err) reject(err);
+      if (data === null)
+        reject(new Error('Response could not be parsed as JSON'));
+      else resolve(data);
     };
-  };
+  }
 
-  prototype.request = function(method, uriParts, opts, callback) {
-    var self = this;
-    opts = opts || {};
-    if (!callback && (typeof opts === 'function')) {
+  request(method, uriParts, opts = {}, callback) {
+    if (!callback && typeof opts === 'function') {
       callback = opts;
       opts = {};
     }
-    if (!callback) {
-      throw "Must supply a callback."
-    }
-    _.assign(opts, {
-      'method': method.toUpperCase(),
-      'uri': self.makeAbsoluteURI(self.makeRelativeURI(uriParts)),
-    });
-    self.addHeaders(opts);
-    request(opts, self.makeRequestCallback(callback));
-  };
+    if (!callback) callback = () => {};
 
-  _.forEach(['get', 'post', 'put', 'delete'], function(method) {
-    prototype[method] = _.partial(prototype.request, method);
-  });
-
-  prototype.getProducts = function(callback) {
-    var self = this;
-    return prototype.get.call(self, ['products'], callback);
-  };
+    Object.assign(opts, {
+      method: method.toUpperCase(),
+      uri: this.makeAbsoluteURI(this.makeRelativeURI(uriParts)),
+    });
+    this.addHeaders(opts);
+    // console.log('making request=', opts);
+    return new Promise((resolve, reject) => {
+      request(opts, this.makeRequestCallback(callback, resolve, reject));
+    });
+  }
 
-  prototype.getProductOrderBook = function(args, callback) {
-    var self = this;
+  getProducts(callback) {
+    return this.get(['products'], callback);
+  }
 
-    args = args || {}
-    if (!callback && (typeof args === 'function')) {
+  getProductOrderBook(args = {}, callback) {
+    if (!callback && typeof args === 'function') {
```

Comment:
In terms of optimizing syntactic aesthetics, decorators are the first thing I think of,

```js
@optionalArgs
someFunction(args, otherParam, callback) {
  /* ... */
}
```

But even though [decorators are in Stage 2](https://tc39.github.io/proposal-decorators/), there isn't great out-of-the-box support for transpiling them quite yet, including in the latest version of [babel](https://babeljs.io/docs/plugins/transform-decorators/).